duffy/api_v2/views.py

```diff
@@ -0,0 +1,206 @@
+        openstack_host.hostname = hostname
+        openstack_host.ip = instance.networks.values()[0][0]
+        openstack_host.state = 'Contextualizing'
+        openstack_host.next_state = 'Active'
+        openstack_host.comment = ''
+        openstack_host.image_source =  instance.image['id']
+        openstack_host.flavovr = flavor
+
+        openstack_host.save()
+    except:
+        print('Error: DB')
+        return jsonify('failed to allocate node to the user'), 412
+
+    sess = Session()
+    sess.apikey = get_key
+    sess.save()
+    if openstack_host.contextualize(project):
+        sess.hosts.append(openstack_host)
+        openstack_host.state = 'Deployed'
+        openstack_host.save()
+    else:
+        # if any of the hosts fail we should return any in-progress hosts
+        # to the pool for reclamation
+        for host in sess.hosts:
+            host.state = 'Active'
+            host.save()
+        sess.state = 'Failed'
+        sess.save()
+        return jsonify('Failed to allocate nodes')
+    sess.save()
+
+    rtn = SessionSchema().dump(sess)
+    return jsonify(rtn.data)
+
+@blueprint.route('/Node/done')
+@duffy_key_required
+@ssid_required
+def nodedone():
+    get_key = request.args.get('key')
+    get_ssid = request.args.get('ssid')
+
+    session = Session.query.get(get_ssid)
+
+    if session.apikey != get_key:
+        return jsonify({'msg': 'Invalid duffy key'}), 403
+
+    if session.state not in ('Prod','Fail'):
+        return jsonify({'msg': 'Session not in Prod'})
+
+    for host in session.hosts:
+        host.state = "Deprovision"
+        host.session = None
+        host.session_id = ''
+        host.save()
+    session.state = 'Done'
+    session.save()
+    return jsonify("Done")
+
+@blueprint.route('/Node/fail')
+@duffy_key_required
+@ssid_required
+def nodefail():
+    get_key = request.args.get('key')
+    get_ssid = request.args.get('ssid')
+
+    session = Session.query.get(get_ssid)
+
+    if session.apikey != get_key:
+        return jsonify({'msg': 'Invalid duffy key'}), 403
+
+    for host in session.hosts:
+        host.state = 'Fail'
+        host.save()
+    session.state = 'Fail'
+    session.save()
+
+    return jsonify("Done")
+
+@blueprint.route('/Inventory')
+def inventory():
+    get_key = request.args.get('key')
+    if get_key:
+        # Return a list of active sessions for the user whose key we have
+        sessions = Session.query.filter(Session.apikey == get_key)
+        rtn_sessions = []
+        for session in sessions:
+            for host in session.hosts:
+                sch = HostSchema().dump(host)
+                rtn_sessions.append([sch.data['hostname'],sch.data['session']])
+        return jsonify(rtn_sessions)
+    else:
+        # No key, return a list of all hosts
+        hosts = Host.query.all()
+        rtn_hosts = []
+
+        for host in hosts:
+            sch = HostSchema().dump(host)
+            ordered_host = [sch.data['id'],
```

Comment:
Let's work out something here. I want to see if we can get away from using the HostSchema from the v2 api on.

The HostSchema was in here because we did that weird thing in Duffy1 that stored the session id in the 'comment' field. Since we don't do that any more, we don't need shouldn't need to filter through a HostSchema any more. Or we can redo a better schema in the future.